routes/user.js

```diff
@@ -0,0 +1,186 @@
+          }
+        } catch (e) {
+          console.warn("Login via login form failed with exception", e)
+          // TODO handle error here
+        }
+      }
+    })
+
+    app.get("/logout", (req, res) => {
+      req.session.ckan_user = false
+      res.redirect("/")
+    })
+
+    app.get("/signup", (req, res, next) => {
+      res.render("user/signup.html")
+    })
+
+    app.post(
+      "/signup",
+      [
+        check("username", "Username is required").isLength(2),
+        check(
+          "fullname",
+          "Full name must be a least 2 characters long"
+        ).isLength(2),
+        check("email", "Please enter a valid email").isLength(3),
+        check("password", "Password field is required")
+          .isLength({ min: 10 })
+          .custom((value, { req, loc, path }) => {
+            if (value !== req.body.confirmPassword) {
+              // trow error if passwords do not match
+              throw new Error("Passwords don't match")
+            } else {
+              return value
+            }
+          })
+      ],
+      async (req, res) => {
+        const validation = validationResult(req)
+        
+        if (!validation.isEmpty()) {
+          const result = validation.errors.reduce((acc, cur) => {
+            const name = cur.param
+            acc[cur.param] = acc[cur.param] || {}
+            acc[name].value = cur.value || req.body[name]
+            acc[name].msg = acc[name].msg
+              ? `${acc[name].msg}, ${cur.msg}`
+              : cur.msg
+            return acc
+          }, {})
+
+          // format submitted vals for form template
+          const allVals = Object.keys(req.body).reduce((acc, key) => {
+            acc[key] = { value: req.body[key] }
+            return acc
+          }, {})
+
+          const r = Object.assign({}, allVals, result)
+
+          return res.render("user/signup.html", r)
+        } else {
+          // if local form validation succeeds, continue to API:
+          const APIResponse = await doAPIUserCreate(req.body)
+
+          // handle errors
+          if (APIResponse.error) {
+            // preserve entered form values
+            const allVals = Object.keys(req.body).reduce((acc, key) => {
+              acc[key] = { value: req.body[key] }
+              return acc
+            }, {})
+
+            const locals = Object.assign({}, allVals, {
+              APIError: APIResponse.error
+            })
+
+            // render signup form with API Error
+            return res.render("user/signup.html", locals)
+          } else {
+            // success!
+            const newUser = APIResponse.result
+            req.session.ckan_user = newUser
+            res.redirect("/profile")
+          }
+        }
+      }
+    )
+
+    app.get("/profile", (req, res) => {
+      if (!req.session.ckan_user) res.redirect("/")
+      res.render("user/profile.html", { user: req.session.ckan_user })
+    })
+
+    app.get("/login-failure", (req, res) => {
```

Comment:
I would prefer not having additional route for this but just showing a flash message on `/login` page.